### PR TITLE
Reply in e2e room

### LINKED ISSUE #211: Handle reply in e2e rooms
There is a trick to do with m.relates data for reply to work in e2e rooms. See riot implem as a reference

### LINKED ISSUE #242: e2e / Error when message contains "
Step to reproduce:
* send a message with quotes in e2e room (hello "test")

-> Result: Message is marked as fail to decrypt

Issue in canonical json

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #211 
Fixes #242 

+ Add reply option in e2e room
+ Fix bug 242
+ Show preview of messages in menu and in text composer preview
+ Fix / crash on long click on event due to dagger migration

### Pull Request Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 16
- [x] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [ ] Pull request updates [CHANGES.md](https://github.com/vector-im/riotX-android/blob/develop/CHANGES.md)
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt [hunk 1]
```diff
@@ -146,21 +147,25 @@ data class Event(
                 val adapter = MoshiProvider.providesMoshi().adapter(Event::class.java)
                 mClearEvent = adapter.fromJsonValue(decryptionResult.clearEvent)
 
-            }
-            mClearEvent?.apply {
-                mSenderCurve25519Key = decryptionResult.senderCurve25519Key
-                mClaimedEd25519Key = decryptionResult.claimedEd25519Key
-                mForwardingCurve25519KeyChain = decryptionResult.forwardingCurve25519KeyChain
-                try {
-                    // Add "m.relates_to" data from e2e event to the unencrypted event
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt]: keep comment

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/util/JsonCanonicalizer.kt [hunk 2]
```diff
@@ -100,7 +100,7 @@ object JsonCanonicalizer {
 
                 return result.toString()
             }
-            is String     -> return "\"" + src.toString() + "\""
+            is String     -> return  JSONObject.quote(src.toString())
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/util/JsonCanonicalizer.kt]: remove unecessary `.toString()` (tested ok)

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/relation/DefaultRelationService.kt [hunk 3]
```diff
@@ -119,31 +116,44 @@ internal class DefaultRelationService @Inject constructor(private val context: C
...
             saveLocalEcho(it)
         } ?: return null
-        val sendContentWorkerParams = SendEventWorker.Params(credentials.userId, roomId, event)
-        val sendWorkData = WorkerParamsFactory.toData(sendContentWorkerParams)
 
+        if (cryptoService.isRoomEncrypted(roomId)) {
+            val encryptWork = createEncryptEventWork(event, listOf("m.relates_to"))
+            val workRequest = createSendEventWork(event)
+            TimelineSendEventWorkCommon.postSequentialWorks(context, roomId, encryptWork, workRequest)
+            return CancelableWork(context, encryptWork.id)
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/relation/DefaultRelationService.kt]: why encryptWork.id and not workRequest.id?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/send/EncryptEventWorker.kt [hunk 4]
```diff
@@ -39,11 +39,15 @@ internal class EncryptEventWorker(context: Context, params: WorkerParameters)
...
+            /**Do not encrypt these keys, keep them as is in encrypted content (e.g. m.relates_to)*/
+            val keepKeys: List<String>? = null
     ) : SessionWorkerParams
 
-    @Inject lateinit var crypto: CryptoService
-    @Inject lateinit var localEchoUpdater: LocalEchoUpdater
+    @Inject
+    lateinit var crypto: CryptoService
+    @Inject
+    lateinit var localEchoUpdater: LocalEchoUpdater
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/send/EncryptEventWorker.kt]: you have not the same format rules that us, let's discuss it IRL

File: vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/action/MessageActionsViewModel.kt [hunk 5]
```diff
@@ -35,21 +37,45 @@ import java.text.SimpleDateFormat
...
-        val showPreview: Boolean = false,
-        val canReact: Boolean = false,
-        val senderAvatarPath: String? = null)
-    : MvRxState {
+        val timelineEvent: TimelineEvent?
+) : MvRxState {
+
+    fun senderName(): String = informationData.memberName?.toString() ?: ""
+
+    fun time(): String? = dateFormat.format(Date(timelineEvent?.root?.originServerTs ?: 0))
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/action/MessageActionsViewModel.kt]: if timelineEvent?.root?.originServerTs is null, we will display an Epoch time, which is not what we want. Maybe return "" in this case

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/relation/DefaultRelationService.kt (hunk 3 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/relation/DefaultRelationService.kt]: same than in send message (agree, it needs some rework)

File: vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/action/MessageActionsViewModel.kt (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/action/MessageActionsViewModel.kt]: done

**Pull request merged**